hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/ITestAbfsInputStreamReadFooter.java

```diff
@@ -0,0 +1,199 @@
+      int fileSize = i * ONE_MB;
+      byte[] fileContent = getRandomBytesArray(fileSize);
+      Path testFilePath = createFileWithContent(fs, fileName, fileContent);
+      seekReadAndTest(fs, testFilePath, fileSize - AbfsInputStream.FOOTER_SIZE,
+          AbfsInputStream.FOOTER_SIZE, fileContent);
+    }
+  }
+
+  private AzureBlobFileSystem getFileSystem(boolean optimizeFooterRead)
+      throws IOException {
+    final AzureBlobFileSystem fs = getFileSystem();
+    fs.getAbfsStore().getAbfsConfiguration()
+        .setOptimizeFooterRead(optimizeFooterRead);
+    fs.getAbfsStore().getAbfsConfiguration()
+        .setReadSmallFilesCompletely(false);
+    return fs;
+  }
+
+  private Path createFileWithContent(FileSystem fs, String fileName,
+      byte[] fileContent) throws IOException {
+    Path testFilePath = path(fileName);
+    try (FSDataOutputStream oStream = fs.create(testFilePath)) {
+      oStream.write(fileContent);
+      oStream.flush();
+    }
+    return testFilePath;
+  }
+
+  private void seekReadAndTest(final FileSystem fs, final Path testFilePath,
+      final int seekPos, final int length, final byte[] fileContent) throws IOException {
+    try (FSDataInputStream iStream = fs.open(testFilePath)) {
+      iStream.seek(seekPos);
+      byte[] buffer = new byte[length];
+      iStream.read(buffer, 0, length);
+      assertSuccessfulRead(fileContent, seekPos, length, buffer);
+      AbfsInputStream abfsInputStream = (AbfsInputStream) iStream
+          .getWrappedStream();
+
+      AzureBlobFileSystem abfs = (AzureBlobFileSystem) fs;
+      AbfsConfiguration conf = abfs.getAbfsStore().getAbfsConfiguration();
+
+      int expectedFCursor = fileContent.length;
+      int expectedLimit;
+      int expectedBCursor;
+      if (conf.optimizeFooterRead()) {
+        expectedBCursor = ((conf.getReadBufferSize() < fileContent.length)
+            ? conf.getReadBufferSize()
+            : fileContent.length);
+        expectedLimit = (conf.getReadBufferSize() < fileContent.length)
+            ? conf.getReadBufferSize()
+            : fileContent.length;
+      } else {
+        expectedBCursor = length;
+        expectedLimit = length;
+      }
+      assertSuccessfulRead(fileContent, abfsInputStream.getBuffer(),
+          conf, length);
+      assertEquals(expectedFCursor, abfsInputStream.getFCursor());
+      assertEquals(expectedBCursor, abfsInputStream.getBCursor());
+      assertEquals(expectedLimit, abfsInputStream.getLimit());
+    }
+  }
+
+  private void assertSuccessfulRead(byte[] actualFileContent,
+      byte[] contentRead, AbfsConfiguration conf, int len) {
+    int buffersize = conf.getReadBufferSize();
+    int actualContentSize = actualFileContent.length;
+    if (conf.optimizeFooterRead()) {
+    len = (actualContentSize < buffersize)
+        ? actualContentSize
+        : buffersize;
+    }
+    assertSuccessfulRead(actualFileContent, actualContentSize - len, len,
```

Comment:
File content asserts are needed on the data that is returned through the client provided buffer as well.

Reply:
Done